File: dnn-providers/hip-kernel-provider/rocke/platform/tests/core/test_arch_target_identity.py
# Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
# SPDX-License-Identifier: MIT

from __future__ import annotations

import os
import subprocess

import pytest

from rocke.core.arch import (
    ArchTarget,
    arch_from_isa,
    base_arch_from_target_id,
    compiler_target_from_target_id,
    known_arches,
    target_id_from_isa,
)


@pytest.mark.parametrize("arch", known_arches())
def test_known_target_ids_normalize_to_themselves(arch: str) -> None:
    assert base_arch_from_target_id(arch) == arch
    assert compiler_target_from_target_id(arch) == arch
    assert arch_from_isa(f"amdgcn-amd-amdhsa--{arch}") == arch


@pytest.mark.parametrize(
    ("target_id", "base_arch", "compiler_target"),
    [
        ("gfx1250-strict", "gfx1250", "gfx1250"),
        ("gfx942:sramecc+:xnack-", "gfx942", "gfx942:sramecc+:xnack-"),
        ("unexpected-target", "unexpected-target", "unexpected-target"),
    ],
)
def test_target_identity_forms(
    target_id: str, base_arch: str, compiler_target: str
) -> None:
    assert target_id_from_isa(target_id) == target_id
    assert arch_from_isa(target_id) == base_arch
    assert base_arch_from_target_id(target_id) == base_arch
    assert compiler_target_from_target_id(target_id) == compiler_target


@pytest.mark.parametrize(
    ("isa", "target_id"),
    [
        ("amdgcn-amd-amdhsa--gfx1250-strict", "gfx1250-strict"),
        ("amdgcn-amd-amdhsa-opencl-gfx1250-strict", "gfx1250-strict"),
        (
            "vendor-prefix-without-triple-fields-gfx942:sramecc+:xnack-",
            "gfx942:sramecc+:xnack-",
        ),
        (
            "gfx-named-prefix-gfx942:sramecc+:xnack-",
            "gfx942:sramecc+:xnack-",
        ),
    ],
)
def test_target_id_extraction_accepts_all_triple_field_forms(
    isa: str, target_id: str
) -> None:
    assert target_id_from_isa(isa) == target_id


def test_arch_target_uses_base_architecture_rows() -> None:
    assert ArchTarget.from_gfx("gfx1250").gfx == "gfx1250"
    with pytest.raises(KeyError, match="unknown gfx target"):
        ArchTarget.from_gfx("gfx1250-strict")


def test_cpp_target_identity_matches_python() -> None:
    """Compare the native CTest executable with Python when a build is supplied."""
    executable = os.environ.get("ROCKE_ARCH_TARGET_TEST_EXE")
    if not executable:
        pytest.skip("set ROCKE_ARCH_TARGET_TEST_EXE to the native target identity test")
    targets = [
        *known_arches(),
        "gfx00a",
        "unexpected-target",
        "",
        "gfx",
        "gfx-",
        "gfx942_bad",
    ]
    inputs = [
        f"{prefix}{target}{profile}{features}"
        for target in targets
        for prefix in ("", "amdgcn-amd-amdhsa--", "gfx-named-prefix-")
        for profile in ("", "-strict")
        for features in ("", ":", ":sramecc+:xnack-", ":unknown+")
    ]
    result = subprocess.run(
        [executable, *inputs], capture_output=True, text=True, check=True, timeout=30
    )
    actual = [tuple(line.split("\t")) for line in result.stdout.splitlines()]
    expected = []
    for isa in inputs:
        target = target_id_from_isa(isa)
        expected.append(
            (
                target,
                base_arch_from_target_id(target),
                compiler_target_from_target_id(target),
                arch_from_isa(isa),
            )
        )
    assert actual == expected

File: dnn-providers/hip-kernel-provider/rocke/platform/tests/core/test_arch_target_identity.cpp
// Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
// SPDX-License-Identifier: MIT
// Host coverage for the target identity contract in test_arch_target_identity.py.
#include "rocke/arch_target.h"

#include <cstdio>
#include <cstring>

using query_fn = const char* (*)(const char*, char*, size_t);
static int failures = 0;

static void check(query_fn query, const char* input, const char* expected)
{
    // Check every capacity through an exact fit and one byte larger, with
    // sentinels proving that truncation and invalid arguments stay in bounds.
    for(size_t cap = 0; cap <= strlen(expected) + 2; ++cap)
    {
        char out[256];
        memset(out, '#', sizeof(out));
        const char* result = query(input, out, cap);
        size_t len = cap == 0 ? 0 : (strlen(expected) < cap ? strlen(expected) : cap - 1);
        if((cap == 0 && result != NULL)
           || (cap != 0 && (result != out || memcmp(out, expected, len) != 0 || out[len] != '\0'))
           || out[cap] != '#')
        {
            fprintf(
                stderr, "FAIL: input '%s', expected '%s', capacity %zu\n", input, expected, cap);
            ++failures;
        }
    }
    char out = '#';
    if(query(NULL, &out, 1) != NULL || out != '#' || query(input, NULL, 1) != NULL)
        ++failures;
}

static void check_target(const char* target, const char* base, const char* compiler)
{
    char isa[256];
    snprintf(isa, sizeof(isa), "amdgcn-amd-amdhsa--%s", target);
    check(rocke_target_id_from_isa, isa, target);
    check(rocke_target_id_from_isa, target, target);
    check(rocke_base_arch_from_target_id, target, base);
    check(rocke_compiler_target_from_target_id, target, compiler);
    check(rocke_arch_from_isa, isa, base);
    check(rocke_arch_from_isa, target, base);
}

int main(int argc, char** argv)
{
    // Query mode lets external Python/C++ comparisons use the real engine.
    if(argc > 1)
    {
        for(int i = 1; i < argc; ++i)
        {
            char target[1024], base[1024], compiler[1024], arch[1024];
            rocke_target_id_from_isa(argv[i], target, sizeof(target));
            rocke_base_arch_from_target_id(target, base, sizeof(base));
            rocke_compiler_target_from_target_id(target, compiler, sizeof(compiler));
            rocke_arch_from_isa(argv[i], arch, sizeof(arch));
            printf("%s\t%s\t%s\t%s\n", target, base, compiler, arch);
        }
        return 0;
    }

    int count = 0;
    const char* const* arches = rocke_known_arches(&count);
    if(count == 0)
        ++failures;
    for(int i = 0; i < count; ++i)
        check_target(arches[i], arches[i], arches[i]);

    check_target("gfx1250-strict", "gfx1250", "gfx1250");
    check_target("gfx942:sramecc+:xnack-", "gfx942", "gfx942:sramecc+:xnack-");
    check_target("gfx1250-strict:xnack-", "gfx1250", "gfx1250:xnack-");
    check_target("gfx11-generic-strict:xnack-", "gfx11-generic", "gfx11-generic:xnack-");
    check_target("gfx00a-profile:unknown+", "gfx00a", "gfx00a:unknown+");
    check_target("gfx942:", "gfx942", "gfx942:");
    check(rocke_arch_from_isa, "unexpected-target", "unexpected-target");
    check(rocke_target_id_from_isa, "unexpected-target", "unexpected-target");
    check(rocke_base_arch_from_target_id, "unexpected-target", "unexpected-target");
    check(rocke_compiler_target_from_target_id, "unexpected-target", "unexpected-target");
    check(rocke_arch_from_isa, "amdgcn-amd-amdhsa-opencl-gfx1250-strict", "gfx1250");
    check(rocke_target_id_from_isa, "gfx-named-prefix-gfx942:xnack-", "gfx942:xnack-");
    check(rocke_arch_from_isa, "gfx-named-prefix-gfx942:xnack-", "gfx942");
    check(rocke_target_id_from_isa, "", "");
    check(rocke_base_arch_from_target_id, "", "");
    check(rocke_compiler_target_from_target_id, "", "");
    check(rocke_arch_from_isa, "", "");
    return failures == 0 ? 0 : 1;
}
